examples/defaulter-gen/generators/defaulter.go

```diff
@@ -899,80 +906,89 @@ func (n *callNode) writeCalls(varName string, isVarPointer bool, sw *generator.S
+		mapDefaultVar := args["index"].(string) + "_default"
 		args["mapDefaultVar"] = mapDefaultVar
-		sw.Do("if reflect.ValueOf($.var$[$.index$]).IsZero() {\n", generator.Args{"var": varName, "index": index})
+	} else {
+		// Defaulting for primitive type
+		variablePlaceholder = "$.varName$"
+	}
 
-		if n.defaultIsPrimitive {
-			if n.defaultDepth > 0 {
-				sw.Do("var ptrVar$.varDepth$ $.varType$ = $.defaultValue$\n", args)
-				for i := n.defaultDepth; i > 0; i-- {
-					sw.Do("ptrVar$.ptri$ := &ptrVar$.i$\n", generator.Args{"i": fmt.Sprintf("%d", i), "ptri": fmt.Sprintf("%d", (i - 1))})
-				}
-				sw.Do("$.varName$[$.index$] = ptrVar0", args)
-			} else {
-				sw.Do("$.varName$[$.index$] = $.defaultValue$", args)
+	variablePointer := variablePlaceholder
+	if !isVarPointer {
+		variablePointer = "&" + variablePointer
+	}
+
+	if n.defaultIsPrimitive {
+		// If the default value is a primitive when the assigned type is a pointer
+		// keep using the address-of operator on the primitive value until the types match
+		if n.defaultDepth > 0 {
+			sw.Do(fmt.Sprintf("if %s == nil {\n", variablePlaceholder), args)
+			sw.Do("var ptrVar$.varDepth$ $.varType$ = $.defaultValue$\n", args)
+			// We iterate until a depth of 1 instead of 0 because the following line
+			// `if $.varName$ == &ptrVar1` accounts for 1 level already
+			for i := n.defaultDepth; i > 1; i-- {
+				sw.Do("ptrVar$.ptri$ := &ptrVar$.i$\n", generator.Args{"i": fmt.Sprintf("%d", i), "ptri": fmt.Sprintf("%d", (i - 1))})
 			}
+			sw.Do(fmt.Sprintf("%s = &ptrVar1", variablePlaceholder), args)
 		} else {
-			sw.Do("$.mapDefaultVar$ := $.varName$[$.index$]\n", args)
-			sw.Do("if err := json.Unmarshal([]byte(`$.defaultValue$`), &$.mapDefaultVar$); err != nil {\n", args)
-			sw.Do("panic(err)\n", nil)
-			sw.Do("}\n", nil)
-			sw.Do("$.varName$[$.index$] = $.mapDefaultVar$\n", args)
+			// For primitive types, nil checks cannot be used and the zero value must be determined
+			sw.Do(fmt.Sprintf("if %s == $.defaultZero$ {\n", variablePlaceholder), args)
+			sw.Do(fmt.Sprintf("%s = $.defaultValue$", variablePlaceholder), args)
 		}
 	} else {
-		sw.Do("if reflect.ValueOf($.var$).IsZero() {\n", generator.Args{"var": varName})
-
-		if n.defaultIsPrimitive {
-			if n.defaultDepth > 0 {
-				sw.Do("var ptrVar$.varDepth$ $.varType$ = $.defaultValue$\n", args)
-				for i := n.defaultDepth; i > 0; i-- {
-					sw.Do("ptrVar$.ptri$ := &ptrVar$.i$\n", generator.Args{"i": fmt.Sprintf("%d", i), "ptri": fmt.Sprintf("%d", (i - 1))})
-				}
-				sw.Do("$.varName$ = ptrVar0", args)
-			} else {
-				sw.Do("$.varName$ = $.defaultValue$", args)
-			}
+		sw.Do(fmt.Sprintf("if %s == nil {\n", variablePlaceholder), args)
+		// Map values are not directly addressable and we need a temporary variablePlaceholder to do json unmarshalling
+		if n.key {
+			sw.Do("$.mapDefaultVar$ := $.varName$[$.index$]\n", args)
+			sw.Do("if err := json.Unmarshal([]byte(`$.defaultValue$`), &$.mapDefaultVar$); err != nil {\n", args)
```

Comment:
so if we had a `map[string]*string`, we would be in that specific condition, right? That's not optimal, but I think that's fine for now. Maybe add a comment?

Reply:
The `*string` is a pointer to a primitive type so it would go down the defaultIsPrimitive code path. This specific condition is reached if we have something like `map[string]SubStruct where the map values are not primitives.